src/tests/cmocka/test_dp_request_table.c

```diff
@@ -0,0 +1,338 @@
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <talloc.h>
+#include <tevent.h>
+#include <errno.h>
+#include <popt.h>
+
+#include "providers/backend.h"
+#include "providers/data_provider/dp_private.h"
+#include "tests/cmocka/common_mock.h"
+#include "tests/common.h"
+
+struct test_ctx {
+    hash_table_t *table;
+};
+
+static int test_setup(void **state)
+{
+    struct test_ctx *test_ctx = NULL;
+    errno_t ret;
+
+    assert_true(leak_check_setup());
+
+    test_ctx = talloc_zero(global_talloc_context, struct test_ctx);
+    assert_non_null(test_ctx);
+
+    ret = dp_req_table_init(test_ctx, &test_ctx->table);
+    assert_int_equal(ret, EOK);
+    assert_non_null(test_ctx->table);
+
+    *state = test_ctx;
+
+    return 0;
+}
+
+static int test_teardown(void **state)
+{
+    talloc_zfree(*state);
+    assert_true(leak_check_teardown());
+    return 0;
+}
+
+static void test_add_del_req(void **state)
+{
+    errno_t ret;
+    bool is_present;
+    const char *req_name;
+    TALLOC_CTX *mem_ctx = global_talloc_context;
+    struct test_ctx *test_ctx = talloc_get_type(*state, struct test_ctx);
+
+    req_name = dp_req_table_key(mem_ctx,
+                                DPT_ID, DPM_ACCOUNT_HANDLER,
+                                DP_FAST_REPLY,
+                                "custom_part");
+    assert_string_equal(req_name, "0:0:0x0001:custom_part");
+
+    is_present = dp_req_table_has_key(test_ctx->table, req_name);
+    assert_false(is_present);
+
+    ret = dp_req_table_add(test_ctx->table, req_name, NULL, NULL);
+    assert_int_equal(ret, EOK);
+
+    is_present = dp_req_table_has_key(test_ctx->table, req_name);
+    assert_true(is_present);
+
+    dp_req_table_del(test_ctx->table, req_name);
+
+    is_present = dp_req_table_has_key(test_ctx->table, req_name);
+    assert_false(is_present);
+
+    talloc_free(discard_const(req_name));
+}
+
+static void test_del_non_present_req(void **state)
+{
+    bool is_present;
+    hash_table_t *table;
+    const char *req_name;
+    TALLOC_CTX *mem_ctx = global_talloc_context;
+    struct test_ctx *test_ctx = talloc_get_type(*state, struct test_ctx);
+
+    table = test_ctx->table;
+
+    req_name = dp_req_table_key(mem_ctx,
+                                DPT_AUTH, DPM_AUTH_HANDLER, 0, "custom_part");
+    assert_string_equal(req_name, "1:1:0000:custom_part");
+
+
+    is_present = dp_req_table_has_key(table, req_name);
+    assert_false(is_present);
+
+    dp_req_table_del(table, req_name);
+
+    is_present = dp_req_table_has_key(table, req_name);
+    assert_false(is_present);
+
+    talloc_free(discard_const(req_name));
+
+    /* Free table */
+    talloc_free(table);
```

Comment:
This belongs to teardown (where it is not necessary thanks to talloc) since you create the table in setup.